### PR TITLE
Fix #9790 - Crash with Pipe:Indoor when temperature is below 1.85C

### LINKED ISSUE #9790: Crash with Pipe:Indoor when temperature is below 1.85C
Issue overview
--------------

Trying to a run a PlantLoop with some Pipe:Indoor, it crashes in CalcPipeHeatTransCoef: https://github.com/NREL/EnergyPlus/blob/5306b934041bf2dd2b2f8e641d79744c6a74a5d3/src/EnergyPlus/PipeHeatTransfer.cc#L1759

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): 22.2.0, develop as well

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added: https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9790/in.idf
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fix #9790

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/PipeHeatTransfer.cc [hunk 1]
```diff
@@ -1748,24 +1747,23 @@ Real64 PipeHTData::CalcPipeHeatTransCoef(EnergyPlusData &state,
...
 
+    if (idx == 0) {
+        PRactual = Pr[idx];
+    } else if (idx >= NumOfPropDivisions) {
+        PRactual = Pr[NumOfPropDivisions - 1];
+    } else {
+        InterpFrac = (Temperature - Temps[idx - 1]) / (Temps[idx] - Temps[idx - 1]);
+        PRactual = Pr[idx - 1] + InterpFrac * (Pr[idx] - Pr[idx - 1]);
+    }
+
```
Comment (jmarrec) [inline src/EnergyPlus/PipeHeatTransfer.cc:1766]: Fix is here. This was an off-by-one error, introduced when modernizing from Array1D to constexpr std::array

File: tst/EnergyPlus/unit/Pipes.unit.cc [hunk 2]
```diff
@@ -77,4 +79,55 @@ TEST_F(EnergyPlusFixture, TestPipesInput)
...
+    EXPECT_EQ(1, pipe.plantLoc.branchNum);
+    EXPECT_EQ(1, pipe.plantLoc.compNum);
+
+    constexpr Real64 massFlowRate = 0.5;
+    constexpr Real64 diameter = 0.05;
+    // Try a temperrature below the min
+    EXPECT_NO_THROW(pipe.CalcPipeHeatTransCoef(*state, 1.0, massFlowRate, diameter));
+    // Try a temperature above the max
+    EXPECT_NO_THROW(pipe.CalcPipeHeatTransCoef(*state, 65.0, massFlowRate, diameter));
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/Pipes.unit.cc:131]: New test. It throws on line 128 before fix

**Comment by jmarrec** [comment]
Defect file on https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9790/in.idf with chicago epw, running in debug mode.

### develop

```cpp
/usr/include/c++/10/debug/array:155:
In function:
    constexpr std::__debug::array<_Tp, _Nm>::value_type& 
    std::__debug::array<_Tp, _Nm>::operator[](std::__debug::array<_Tp, 
    _Nm>::size_type) [with _Tp = double; long unsigned int _Nm = 20; 
    std::__debug::array<_Tp, _Nm>::reference = double&; 
    std::__debug::array<_Tp, _Nm>::value_type = double; 
    std::__debug::array<_Tp, _Nm>::size_type = long unsigned int]

Error: attempt to subscript container with out-of-bounds index 20, but 
container only holds 20 elements.

Objects involved in the operation:
    sequence "this" @ 0x0x558a58cf94a0 {
      type = std::__debug::array<double, 20ul>;
    }
Aborted (core dumped
```

```gdb
(lldb) bt
* thread #1, name = 'energyplus', stop reason = signal SIGABRT
  * frame #0: 0x00007fffe5b2200b libc.so.6`__GI_raise(sig=<unavailable>) at raise.c:51:1
    frame #1: 0x00007fffe5b01859 libc.so.6`__GI_abort at abort.c:79:7
    frame #2: 0x00007fffe5db3179 libstdc++.so.6`___lldb_unnamed_symbol7383 + 5
    frame #3: 0x00007fffecabba4b libenergyplusapi.so.23.1.0`std::__debug::array<double, 13ul>::_Array_check_subscript<13ul>::_Array_check_subscript(this=0x0000555556679d90, __index=18446744073709551615) at array:68:6
    frame #4: 0x00007fffecabb72e libenergyplusapi.so.23.1.0`std::__debug::array<double, 13ul>::operator[](this=0x00007ffff11546c0, __n=18446744073709551615) const at array:163:6
    frame #5: 0x00007fffeea5a7d5 libenergyplusapi.so.23.1.0`EnergyPlus::PipeHeatTransfer::PipeHTData::CalcPipeHeatTransCoef(this=0x000055555654d280, state=0x00007fffffffafc0, Temperature=1.0090331019641647, MassFlowRate=4.2776686290037569, Diameter=0.050000000000000003) at PipeHeatTransfer.cc:1759:58
    frame #6: 0x00007fffeea56459 libenergyplusapi.so.23.1.0`EnergyPlus::PipeHeatTransfer::PipeHTData::CalcPipesHeatTransfer(this=0x000055555654d280, state=0x00007fffffffafc0, LengthIndex=ObjexxFCL::Optional_int_const @ 0x00007fffffff9a40) at PipeHeatTransfer.cc:1224:48
    frame #7: 0x00007fffeea41017 libenergyplusapi.so.23.1.0`EnergyPlus::PipeHeatTransfer::PipeHTData::simulate(this=0x000055555654d280, state=0x00007fffffffafc0, calledFromLocation=0x0000555556ba7290, FirstHVACIteration=false, CurLoad=0x0000555556ba71a0, RunFlag=true) at PipeHeatTransfer.cc:161:40
```

### With fix

Simulation completes successfully.

**Review by Myoldmopar — APPROVED** [review]
This looks like a good fix for the issue.  Thanks @jmarrec

**Comment by Myoldmopar** [comment]
Tested locally and everything is happy, merging.

**Pull request merged**